[dtensor] remove assertions about submesh checks

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101229
* #101202

This PR removes assertions from submesh checks to directly return local
tensor, this is so that all the other APIs can work with submesh

wanchaol committed:
[dtensor] remove assertions about submesh checks

This PR removes assertions from submesh checks to directly return local
tensor, this is so that all the other APIs can work with submesh

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -184,11 +184,11 @@ def _reduce_shard_tensor(
-            my_coordinate is not None
-        ), "Rank if not part of mesh"  # TODO: figure out behavior here
+
+        if my_coordinate is None:
+            # if rank is not part of mesh, we simply return local_tensor,
+            # which should be an empty tensor
+            return tensor
```
So this is essentially a no-op right?

wanchaol replied:
yes, it's a no-op if rank is not part of mesh

wanchaol committed:
Update on "[dtensor] remove assertions about submesh checks"


This PR removes assertions from submesh checks to directly return local
tensor, this is so that all the other APIs can work with submesh

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
